TM/Tutor Follow Evolutions

AaronCFreytag opened the pull request:
This is a feature I added for my personal fork and I thought it might be something that could be appropriate for the main repo.

These settings cause TM/Tutor compatibility to carry up evolution lines (including split evolutions) with a chance of gaining additional compatibility with previously incompatible TMs/Tutors. Flat 25% chance per evolution normally, or with "prefer same type" enabled, 90% on newly gained types and 10% on everything else.

This change also refactors the TM/Tutor compatibility parts of the program slightly to use the same method for both. I did a few tests to make sure it still gives the same result as before when using the same seed and settings. There are also some debug methods for logging TM/Tutor compatibility which aren't used currently but I left them in since they might be useful in the future for checking things.

I ended up moving a couple things around in the UI, putting the options specifically related to TM compatibility and move tutor compatibility in the compatibility sections with the radio buttons (mostly because I needed a little extra room for the new checkboxes). The new options are also saved in the settings string.

Changed UI:
![image](https://user-images.githubusercontent.com/15931642/109253428-a4e8e380-77a4-11eb-9167-27e0d9663e7c.png)

I know this is kind of out of the blue for a significant change but since I had already made the feature on my own I thought I would just shoot out a PR and see if it would be something that would be of interest for the main repo.

Thanks!

tom-overton commented:
Can you supply screenshots of the new UI? I'm giving the code a look right now but it's hard to tell what's different just looking at changes to the .form file.

AaronCFreytag commented:
Updated the initial post with a screenshot of the UI change. I mostly just changed it like this to reasonably fit the "follow evolutions" checkboxes since I thought it would be a bit cramped on the left side, but I can change it back and find somewhere else to fit the checkboxes potentially.

tom-overton commented:
Thanks. The change looks totally sensible in the screenshot; I guess pictures really are worth a thousand words. Outside of code stuff, I have a couple of questions about the implications of this feature I just want to make sure we're all on the same page about:
1. Unlike all other "follow evolution" settings, this one seems to carry forward the TM/Move Tutor Compatibility to *all* branches of a split evo. In my research, this actually seems to line up with how things work in the vanilla game; I haven't found a single case where a Pokemon is compatible with a TM or move tutor but then loses that compatibility when it evolves. This results in weird cases where Vaporeon can learn Sunny Day and Flareon can learn Rain Dance, for example. So this behavior is inconsistent with all other "follow evolution" settings but is consistent with how the original game works.
2. Assuming someone *doesn't* use the "Limit to Three Stages" setting for evolutions, the max evolution chain length is 10. Following the logic of "all previous TMs/Tutors copy over, and there's a chance of getting more of them," a really long evolution chain has a decent chance of making the final stage compatible with almost every TM/Tutor in the game.

I don't think either of these are problems per se; they're just somewhat weird/inconsistent things that I wanted to be crystal clear were intentional.

tom-overton commented on src/com/dabomstew/pkrandom/newgui/Bundle.properties:
```diff
@@ -574,4 +574,8 @@ GUI.settingsStringTooNew=The settings string was generated with a newer randomiz
\ No newline at end of file
+GUI.pleaseUseTheLauncher=<html>The randomizer was not started via the launcher. You will not be able to load 3DS games into the randomizer.<br /><br />If you are seeing this message despite using the launcher, make sure that you are using the launcher bundled<br /> with version 4.0.2 (or later) of the randomizer.
+GUI.tmFollowEvolutionsCheckBox.text=Follow Evolutions
+GUI.tmFollowEvolutionsCheckBox.toolTipText=Causes TM/HM compatibility to follow along with evolutions.
```
This is minor but it would be nice to see these mirror something like `GUI.paFollowEvolutionsCheckBox.toolTipText`. So it would look something like "<html>When this is selected and TM compatibility is randomized...."
You could even insert line breaks with and explain how you gain more TMs/Tutors when evolving, and how the odds are this are affected by whether or not you prefer same type.

AaronCFreytag replied:
How does this look?
![image](https://user-images.githubusercontent.com/15931642/109280024-5270ec80-77cf-11eb-9e8c-721209651456.png)
![image](https://user-images.githubusercontent.com/15931642/109280037-56047380-77cf-11eb-948b-65da068fc174.png)

tom-overton replied:
Looks good to me but I'm not the best with strings. I'll leave this unresolved for now so Ajarmar to take a look.

Ajarmar replied:
Yeah, looks good

tom-overton reviewed (commented):
Looks mostly good to me, and the feature seems pretty cool. I'll make sure Ajarmar takes a look to make sure we're both okay with the feature.

AaronCFreytag commented:
Thanks for checking out the PR! To answer your questions:

   1. This was a design choice I made when implementing the feature to avoid split evolution pokemon losing compatibility with previous TMs/Tutors when evolving. This does mean that the difference between split evos will be less drastic compared to how it is without having "follow evolutions" enabled, but I think it's a good tradeoff for the QOL provided by not having to worry about pokemon losing TM/tutor compatibility when evolving, and as you mentioned it does work this way in the original pokemon games so I think it makes sense.
 
   2. This was something I had considered, although mostly in the context of 1-3 evolution chain pokemon. I had decided to leave it at a flat chance per evolution just for simplicity's sake, but a 10-stage evolution pokemon with completely random compatibilty would have around a 95% chance of being compatible with any given TM which would be pretty crazy. I'm not sure how frequently 10-stage evolution chain pokemon come up in randomization, but if it seems too ridiculous I could reduce the probability of learning moves through evolution, or potentially do something a little more complex like scaling the probability based on the number of evolution stages

tom-overton commented:
1. Yeah I figured that was a deliberate design choice. I'm fine with the inconsistency here because, like you said, it's nice QOL, and the original game works like this too.
2. I doubt a chain that long occurs often in practice. I wouldn't change the current probabilities since they seem totally reasonable for smaller evolution lines, and I don't think this PR needs to add the complexity of scaling the probability either. If people *actually* see this in practice and bring it up to us, a follow-up change can be made to scale the probability, but since this scenario is so rare I doubt someone will even run into it.

AaronCFreytag commented on src/com/dabomstew/pkrandom/Settings.java:
```diff
@@ -49,7 +49,7 @@
     public static final int VERSION = Version.VERSION;
 
-    public static final int LENGTH_OF_SETTINGS_DATA = 48;
+    public static final int LENGTH_OF_SETTINGS_DATA = 49;
```
I actually originally had the TMs setting here but I wasn't sure if it would be appropriate so I moved it to the end. At any rate, I went ahead and moved the TM and tutor evolution settings to byte 19 and reduced the settings length back to 48.

Ajarmar commented on src/com/dabomstew/pkrandom/RomFunctions.java:
```diff
@@ -39,22 +39,22 @@
+    public static Set<Pokemon> getBasicPokemon(RomHandler baseRom, boolean includeNonCopyingMons) {
         List<Pokemon> allPokes = baseRom.getPokemonInclFormes();
-        Set<Pokemon> dontCopyPokes = new TreeSet<>();
+        Set<Pokemon> basicPokemon = new TreeSet<>();
```
I'm not sure about the method/variable renaming here. This method is only used in `copyUpEvolutionsHelper`, where its return value is set to a variable named `dontCopyPokes`, and the point of the method is explicitly to find the Pokemon for which something should not be copied (which includes basic Pokemon, and in other cases also split evolutions). In other words, the focus of the method is to find "No Copy Pokemon", not "Basic Pokemon". I'd like to have the method name and the name of this Set changed back, and the new input parameter could just be named `dontCopySplitEvos`.

AaronCFreytag replied:
Changed it back. I mostly just renamed it since I thought I could shorten it by having the copying part in the parameters, but based on the usage it does make a lot of sense the way it was before.

Also, looking at this again, I realize that `getMiddleEvolutions` and `getFinalEvolutions` also account for split evolutions in the same way. I'll add the parameter to these methods too, mostly so I can fix the `toMonIsFinalEvo` boolean in case it needs to be used in the future with `dontCopySplitEvos` enabled.

Ajarmar commented on src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:
```diff
@@ -4658,8 +4737,11 @@ private boolean isCyclic(Pokemon pk, Set<Pokemon> visited, Set<Pokemon> recStack
      */
-    private void copyUpEvolutionsHelper(BasePokemonAction bpAction, EvolvedPokemonAction epAction) {
+    private void copyUpEvolutionsHelper(BasePokemonAction bpAction, EvolvedPokemonAction epAction,
+                                        boolean ignoreNoCopyFlag) {
```
Just in the interest of making this a bit less confusing, maybe this new input parameter also should be named `dontCopySplitEvos` and not be inverted when it's used?

Ajarmar reviewed (changes requested):
Hey, this seems like a great feature! Nice work on this. I have just a few things that I think should be changed.

AaronCFreytag commented on src/com/dabomstew/pkrandom/newgui/NewRandomizerGUI.java:
```diff
@@ -3088,9 +3104,12 @@ private void enableOrDisableSubControls() {
             if (!(pmsUnchangedRadioButton.isSelected()) || !(tmUnchangedRadioButton.isSelected())
                     || !(thcUnchangedRadioButton.isSelected())) {
                 tmLevelupMoveSanityCheckBox.setEnabled(true);
+                tmFollowEvolutionsCheckBox.setEnabled(true);
```
Ahh yeah, that makes sense. Done adding that, I also moved it out of the level up sanity conditional branch just so disabling the checkbox only needs to be in one place.

tom-overton reviewed (approved):
I played around with this and it seems to work well, so with the changes you've added, everything looks good to me. Once Ajarmar approves as well, we can merge this.

Ajarmar commented:
Sorry to ask for more changes, but I'm still not entirely satisfied with the variable naming for the "dontCopy" stuff. I could've been a bit clearer with what I meant.

If `dontCopySplitEvos` is _true_ in `copyUpEvolutionsHelper`, then the following should happen:

- `getBasicOrNoCopyPokemon` should add split evolutions to its Set.
- `getMiddleEvolutions` should _not_ add split evolutions to its Set.

Currently, the opposite happens instead. To make the variable naming make sense, I'd do the following:

- In `getBasicOrNoCopyPokemon`, in the conditional (line 49), `dontCopySplitEvos` should not be inverted.
- In `copyUpEvolutionsHelper`, when calling `getMiddleEvolutions`, `dontCopySplitEvos` should be inverted.
- When `copyUpEvolutionsHelper` is called in the TM/Tutor methods, the `dontCopySplitEvos` parameter should be false rather than true. (Because you _do_ want to copy split evos in those cases)
- Likewise, when `copyUpEvolutionsHelper` is called in other cases, it should be true rather than false. (Because you don't want to copy split evos in those cases)

Hope this makes it clear what my intention was, but please ask if anything's unclear.

AaronCFreytag commented:
Ohh I see! Sorry for that, my interpretation on the naming was backwards.

Just to make completely sure I've got it right this time:
- In `copyUpEvolutionsHelper`, when `dontCopySplitEvos`, is true, split evolutions will be treated as a base pokemon instead of being treated as an evolution. So for previously existing usages of the function, `dontCopySplitEvos` will be true, but for the special case of TM/Tutor evolution following, `dontCopySplitEvos` will be false.
- In `getBasicOrNoCopyPokemon`, when `dontCopySplitEvos` is true, split evolutions will be included in the returned set alongside first-stage pokemon. In other words, this parameter should behave inverted from how it is currently. When called within `copyUpEvolutionsHelper`, `dontCopySplitEvos` will just be passed directly to it.
- In `getMiddleEvolutions`, when `includeSplitEvos` is true, split evolutions will be included in the returned set alongside other second-stage pokemon when the split evolution is also second-stage. This method by itself is unchanged from before. When called within `copyUpEvolutionsHelper`, `dontCopySplitEvos` will be passed in inverted.

Sorry if that was a bit verbose. I know a lot of this is exactly what you said, but I just wanted to make sure I have things understood correctly now. I'll update the PR with the changes in just a moment.

AaronCFreytag commented:
Don't mind the force push; I just wanted to change the commit message quick

Ajarmar commented:
Yeah, that all seems right. Maybe it's a bit confusing in general to have a boolean with the word "dont" in the name, but we'll keep it like this for now. Everything looks good now, so I'll merge it in. Thanks!
